docs/review.md

```diff
@@ -0,0 +1,579 @@
+
+### Need unit tests for datamapper
+
+TODO I think we should encourage the use of a [shoulda](https://github.com/thoughtbot/shoulda-matchers) style syntax for datamapper:
+
+https://github.com/greyblake/dm-rspec
+
+But I realise that this might not be popular - but I think it's an important part of following the acceptance test, unit test cycle ...
+
+Maybe this should be part of moving to something like:
+
+* https://github.com/jeremyevans/sequel
+* https://github.com/kematzy/minitest-sequel
+* https://github.com/openhood/rspec_sequel_matchers
+
+I guess we can't have any of the above when the bookmark_manager project doesn't address any of this ...
+
+### Appropriate use of Spec Helpers
+
+```ruby
+require_relative 'helpers'
+
+feature "User sign up" do
+  include Helpers
+  let(:user) { user = build(:user) }
+
+  scenario 'with correct credentials' do
+    user = create(:user)
+    sign_in(user)
+    expect(page).to have_content("Welcome, #{user.email}")
+  end
+end
+```
+
+```ruby
+# spec/helpers.rb
+module Helpers
+
+  def sign_in(user)
+    visit('/sessions/new')
+    fill_in :email, with: user.email
+    fill_in :password, with: user.password
+    click_button "Log in"
+  end
+```
+
+See https://github.com/makersacademy/course/blob/master/pills/spec_helper_methods.md
+
+
+## Ensure asset routes are set correctly
+
+Do we like http://recipes.sinatrarb.com/p/asset_management/sinatra_assetpack ?
+
+Or should we just be telling them to set the public folder correctly
+
+```ruby
+set :public_folder, Proc.new { File.join(root, 'static') }
+```
+
+and for Heroku:
+
+* http://abernardes.github.io/2014/10/24/static-assets-with-rack.html
+* http://stackoverflow.com/a/5509152/316729
+
+add the following to config.ru
+
+```ruby
+use Rack::Static, :urls => ['/stylesheets', '/javascripts'], :root => 'public'
+```
+
+# Step 4: Application code and \*.rb files
+
+## Controllers
+
+### Prefer Sinatra::Application vs Sinatra::Base
+
+TODO work out if we care
+
+* http://www.sinatrarb.com/intro.html#Modular%20vs.%20Classic%20Style
+
+### Be clear about how to use flash.now
+
+Have to register Sinatra Flash before using. Have to use
+
+```
+flash.next[:notice]
+```
+
+and not
+
+```
+flash.now[:notice]
+```
+
+in order to see the message after the redirect.
+
+### Refactor long controller methods
+
+This overly long controller method with business logic spread spread out throughout the method is not ideal.
+
+```ruby
+post '/password_reset' do
+  user = User.first(email: params[:email])
+  if user
+    user.reset_password
+    user.save
+    flash[:info] = 'Check your emails'
+    redirect('/')
+  else
+    flash[:info] = 'No such account on record'
+    redirect('/password_reset')
+  end
+end
+```
+
+It could be refactored like so:
+
+```ruby
+post '/password_reset' do
+  if User.reset_password(email: params[:email])
```

Comment:
I'm a big fan of this, although is this done in the curriculum? Could still be something to help them out